lib/popcorn/popcorn.nit

```diff
@@ -48,12 +48,19 @@ redef class App
 	redef fun answer(req, uri) do
 		uri = uri.simplify_path
 		var res = new HttpResponse(404)
+		for route, handler in pre_handlers do
+			handler.handle(route, uri, req, res)
+		end
 		for route, handler in handlers do
 			handler.handle(route, uri, req, res)
+			if res.sent then break
 		end
 		if not res.sent then
 			res.send(error_tpl(res.status_code, res.status_message), 404)
 		end
+		for route, handler in post_handlers do
+			handler.handle(route, uri, req, res)
+		end
```

Comment:
I think that you should clearly document the `sending` policy. If a understand your implementation, it is

* a pre_route should not try send the response, as it will likely break things
* a normal route can send a response, if it does so, then is it the last normal route executed
* if no normal route send the response then is is automatically send after all normal routes. In fact is the is expected pattern for most cases
* after the response id send (automatically ou manually) the all post_routes are executed. Obviously they cannot send the response since it is already gone.